adapter/prometheus2/prometheus.go

```diff
@@ -0,0 +1,287 @@
+// Copyright 2017 Istio Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package prometheus publishes metric values collected by Mixer for
+// ingestion by prometheus.
+package prometheus
+
+import (
+	"context"
+	"fmt"
+	"math"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	multierror "github.com/hashicorp/go-multierror"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promhttp"
+
+	"istio.io/mixer/adapter/prometheus2/config"
+	"istio.io/mixer/pkg/adapter"
+	"istio.io/mixer/template/metric"
+)
+
+type (
+	builder struct {
+		srv      server
+		registry *prometheus.Registry
+		once     sync.Once
+	}
+
+	handler struct {
+		metrics map[string]prometheus.Collector
+		kinds   map[string]config.Params_MetricInfo_Kind
+	}
+)
+
+var (
+	charReplacer = strings.NewReplacer("/", "_", ".", "_", " ", "_")
+
+	_ metric.HandlerBuilder = &builder{}
+	_ metric.Handler        = &handler{}
+)
+
+// GetBuilderInfo returns the BuilderInfo associated with this adapter implementation.
+func GetBuilderInfo() adapter.BuilderInfo {
+	return adapter.BuilderInfo{
+		Name:        "prometheus",
+		Description: "Publishes prometheus metrics",
+		SupportedTemplates: []string{
+			metric.TemplateName,
+		},
+		CreateHandlerBuilder: func() adapter.HandlerBuilder { return newFactory(newServer(defaultAddr)) },
+		DefaultConfig:        &config.Params{},
+		ValidateConfig:       func(msg adapter.Config) *adapter.ConfigErrors { return nil },
+	}
+}
+
+func newFactory(srv server) *builder {
+	return &builder{srv, prometheus.NewPedanticRegistry(), sync.Once{}}
+}
+
+func (b *builder) ConfigureMetricHandler(map[string]*metric.Type) error { return nil }
+
+func (b *builder) Build(c adapter.Config, env adapter.Env) (adapter.Handler, error) {
+	var serverErr error
+	b.once.Do(func() {
```

Comment:
Why is this needed? Can't this just be started eagerly? If not, then a comment would be good to explain why this is so.